File: ios/Plugin/DoNotDisturbPlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(DoNotDisturbPlugin, "DoNotDisturb",
           CAP_PLUGIN_METHOD(monitor, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(set, CAPPluginReturnPromise);
)

File: ios/Plugin/DoNotDisturbPlugin.swift
import Foundation
import Capacitor

@objc(DoNotDisturbPlugin)
public class DoNotDisturbPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DoNotDisturbPlugin"
    public let jsName = "DoNotDisturb"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setEnabled", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = DoNotDisturb()

    public override func load() {
        implementation.startListening { [weak self] in
            guard let self = self else { return }
            self.implementation.isEnabled { enabled in
                self.notifyListeners("dndStateChanged", data: ["enabled": enabled])
            }
        }
    }

    @objc func isEnabled(_ call: CAPPluginCall) {
        implementation.isEnabled { enabled in
            call.resolve(["enabled": enabled])
        }
    }

    @objc func setEnabled(_ call: CAPPluginCall) {
        call.reject("iOS does not allow programmatic DND control")
    }

    deinit {
        implementation.stopListening()
    }
}
